animation_nodes/math/rotation_conversion.pyx

```diff
@@ -122,3 +130,132 @@ cdef void normalizedAxisCosAngleToMatrix(Matrix3_or_Matrix4* m, Vector3* axis, f
     if Matrix3_or_Matrix4 is Matrix4:
         m.a14, m.a24, m.a34 = 0, 0, 0
         m.a41, m.a42, m.a43, m.a44 = 0, 0, 0, 1
+
+# Euler to Quaternion
+##########################################################    
+        
+# https://en.wikipedia.org/wiki/Conversion_between_quaternions_and_Euler_angles
+cdef void euler3ToQuaternion(Quaternion *q, Euler3 *e):
+    cdef float cx = cos(e.x * 0.5)
+    cdef float sx = sin(e.x * 0.5)
+    cdef float cy = cos(e.y * 0.5)
+    cdef float sy = sin(e.y * 0.5)
+    cdef float cz = cos(e.z * 0.5)
+    cdef float sz = sin(e.z * 0.5)
+    
+    cdef float cxcy = cx * cy
+    cdef float sxcy = sx * cy
+    cdef float sxsy = sx * sy
+    cdef float cxsy = cx * sy
+    
+    q.w = cxcy * cz + sxsy * sz
+    q.x = sxcy * cz - cxsy * sz
+    q.y = cxsy * cz + sxcy * sz
+    q.z = cxcy * sz - sxsy * cz
+    
+# Quaternion to Euler
+##########################################################
+
+# https://en.wikipedia.org/wiki/Conversion_between_quaternions_and_Euler_angles
+cdef void quaternionToEuler3(Euler3 *e, Quaternion *q):
+    cdef float sinrCosp = 2 * (q.w * q.x + q.y * q.z)
+    cdef float cosrCosp = 1 - 2 * (q.x * q.x + q.y * q.y)
+    e.x = atan2(sinrCosp, cosrCosp)
+
+    cdef float sinp = 2 * (q.w * q.y - q.z * q.x)
+    if abs(sinp) >= 1.0:
+        e.y = copysign(PI / 2, sinp)
+    else:
+        e.y = asin(sinp)
+
+    cdef float sinyCosp = 2 * (q.w * q.z + q.x * q.y)
+    cdef float cosyCosp = 1 - 2 * (q.y * q.y + q.z * q.z)
+    e.z = atan2(sinyCosp, cosyCosp)
+    e.order = 0
+
+# Quaternion to Matrix
+##########################################################
+
+# https://www.euclideanspace.com/maths/geometry/rotations/conversions/quaternionToMatrix/index.htm
+
+@cython.cdivision(True)
+cdef void quaternionToMatrix4(Matrix4 *m, Quaternion *q):
+    quaternionNormalize_InPlace(q)
+    cdef float w, x, y, z
+    w, x, y, z = q.w, q.x, q.y, q.z
+    cdef float ww = w * w
+    cdef float xx = x * x
+    cdef float yy = y * y
+    cdef float zz = z * z
+
+    cdef float xy = x * y
+    cdef float xz = x * z
+    cdef float yz = y * z
+    cdef float zw = z * w
+    cdef float yw = y * w
+    cdef float xw = x * w
+
+    cdef invs = xx + yy + zz + ww
+
+    m.a11 = (xx - yy - zz + ww) / invs
+    m.a22 = (-xx + yy - zz + ww) / invs
+    m.a33 = (-xx - yy + zz + ww) / invs
+
+    m.a21 = 2.0 * (xy + zw) / invs
+    m.a12 = 2.0 * (xy - zw) / invs
+
+    m.a31 = 2.0 * (xz - yw) / invs
+    m.a13 = 2.0 * (xz + yw) / invs
+
+    m.a32 = 2.0 * (yz + xw) / invs
+    m.a23 = 2.0 * (yz - xw) / invs
```

Comment:
You are still not handling zero division here. Keep it as it was with `1 / (xx + yy + zz + ww)` but first check if `xx + yy + zz + ww` is zero, and if so, return a unit matrix.

```
cdef float dot = xx + yy + zz + ww
if dot == 0:
    setIdentityMatrix(q)
    return
cdef float invs = 1 / dot
```
Notice the `float` after cdef as well, which you seem to forgot.